File: app/src/test/java/com/clockmods/weather/QWeatherSignerTest.java
package com.clockmods.weather;

import net.i2p.crypto.eddsa.EdDSAPrivateKey;
import net.i2p.crypto.eddsa.EdDSAPublicKey;
import net.i2p.crypto.eddsa.EdDSASecurityProvider;
import net.i2p.crypto.eddsa.EdDSAEngine;
import net.i2p.crypto.eddsa.spec.EdDSANamedCurveTable;
import net.i2p.crypto.eddsa.spec.EdDSAParameterSpec;
import net.i2p.crypto.eddsa.spec.EdDSAPrivateKeySpec;
import net.i2p.crypto.eddsa.spec.EdDSAPublicKeySpec;

import org.json.JSONObject;
import org.junit.Assert;
import org.junit.Test;

import java.nio.charset.StandardCharsets;
import java.security.Security;

import org.apache.commons.codec.binary.Base64;

public class QWeatherSignerTest {
    @Test
    public void createsVerifiableQWeatherJwt() throws Exception {
        if (Security.getProvider(EdDSASecurityProvider.PROVIDER_NAME) == null) {
            Security.addProvider(new EdDSASecurityProvider());
        }
        byte[] seed = new byte[32];
        for (int index = 0; index < seed.length; index++) seed[index] = (byte) index;
        EdDSAParameterSpec spec = EdDSANamedCurveTable.getByName(EdDSANamedCurveTable.ED_25519);
        EdDSAPrivateKey privateKey = new EdDSAPrivateKey(new EdDSAPrivateKeySpec(seed, spec));
        String key = Base64.encodeBase64String(privateKey.getEncoded());
        String token = QWeatherSigner.token("CREDENTIAL", "PROJECT", key, 1000L);
        String[] parts = token.split("\\.");
        Assert.assertEquals(3, parts.length);
        JSONObject header = json(parts[0]);
        JSONObject payload = json(parts[1]);
        Assert.assertEquals("EdDSA", header.getString("alg"));
        Assert.assertEquals("CREDENTIAL", header.getString("kid"));
        Assert.assertEquals("PROJECT", payload.getString("sub"));
        Assert.assertEquals(970L, payload.getLong("iat"));
        Assert.assertEquals(1870L, payload.getLong("exp"));
        Assert.assertFalse(token.contains("="));

        EdDSAEngine verifier = new EdDSAEngine();
        verifier.initVerify(new EdDSAPublicKey(new EdDSAPublicKeySpec(privateKey.getAbyte(), spec)));
        verifier.update((parts[0] + "." + parts[1]).getBytes(StandardCharsets.UTF_8));
        Assert.assertTrue(verifier.verify(Base64.decodeBase64(parts[2])));
    }

    private static JSONObject json(String encoded) throws Exception {
        return new JSONObject(new String(Base64.decodeBase64(encoded), StandardCharsets.UTF_8));
    }
}

File: app/src/test/java/com/clockmods/weather/QWeatherSignerTest.kt
package com.clockmods.weather

import java.nio.charset.StandardCharsets
import java.security.Security
import net.i2p.crypto.eddsa.EdDSAPrivateKey
import net.i2p.crypto.eddsa.EdDSAPublicKey
import net.i2p.crypto.eddsa.EdDSASecurityProvider
import net.i2p.crypto.eddsa.EdDSAEngine
import net.i2p.crypto.eddsa.spec.EdDSANamedCurveTable
import net.i2p.crypto.eddsa.spec.EdDSAParameterSpec
import net.i2p.crypto.eddsa.spec.EdDSAPrivateKeySpec
import net.i2p.crypto.eddsa.spec.EdDSAPublicKeySpec
import org.apache.commons.codec.binary.Base64
import org.json.JSONObject
import org.junit.Assert
import org.junit.Test

class QWeatherSignerTest {
    @Test
    fun createsVerifiableQWeatherJwt() {
        if (Security.getProvider(EdDSASecurityProvider.PROVIDER_NAME) == null) Security.addProvider(EdDSASecurityProvider())
        val seed = ByteArray(32) { it.toByte() }
        val spec: EdDSAParameterSpec = EdDSANamedCurveTable.getByName(EdDSANamedCurveTable.ED_25519)
        val privateKey = EdDSAPrivateKey(EdDSAPrivateKeySpec(seed, spec))
        val key = Base64.encodeBase64String(privateKey.encoded)
        val token = QWeatherSigner.token("CREDENTIAL", "DEVELOPER", "PROJECT", key, 1000L)
        val parts = token.split(".")
        Assert.assertEquals(3, parts.size)
        val header = json(parts[0])
        val payload = json(parts[1])
        Assert.assertEquals("EdDSA", header.getString("alg"))
        Assert.assertEquals("CREDENTIAL", header.getString("kid"))
        Assert.assertEquals("DEVELOPER", payload.getString("iss"))
        Assert.assertEquals("PROJECT", payload.getString("sub"))
        Assert.assertEquals(970L, payload.getLong("iat"))
        Assert.assertEquals(1870L, payload.getLong("exp"))
        Assert.assertFalse(token.contains("="))
        val verifier = EdDSAEngine()
        verifier.initVerify(EdDSAPublicKey(EdDSAPublicKeySpec(privateKey.abyte, spec)))
        verifier.update("${parts[0]}.${parts[1]}".toByteArray(StandardCharsets.UTF_8))
        Assert.assertTrue(verifier.verify(Base64.decodeBase64(parts[2])))
    }

    private fun json(encoded: String): JSONObject = JSONObject(String(Base64.decodeBase64(encoded), StandardCharsets.UTF_8))
}
